core/app/models/spree/payment.rb

```diff
@@ -195,9 +195,20 @@ def has_invalid_state?
 
     def validate_source
       if source && !source.valid?
-        source.errors.each do |field, error|
-          field_name = I18n.t("activerecord.attributes.#{source.class.to_s.underscore}.#{field}")
-          errors.add(Spree.t(source.class.to_s.demodulize.underscore), "#{field_name} #{error}")
+        if Rails::VERSION::STRING >= '6.1'
+          source.errors.map { |error| { field: error.attribute, message: error&.message } }.each do |err|
+            next if err[:field].blank? || err[:message].blank?
+
+            field_name = I18n.t("activerecord.attributes.#{source.class.to_s.underscore}.#{err[:field]}")
+            errors.add(Spree.t(source.class.to_s.demodulize.underscore), "#{field_name} #{err[:message]}")
+          end
+        else
+          source.errors.messages.each do |field, error|
+            next if field.blank? || error.empty?
+
+            field_name = I18n.t("activerecord.attributes.#{source.class.to_s.underscore}.#{field}")
+            errors.add(Spree.t(source.class.to_s.demodulize.underscore), "#{field_name} #{error.first}")
```

Comment:
@przemosk please extract this into a new method called `add_source_error(field, message)`